### PR TITLE
notebooks, scripts and models

### PR DESCRIPTION
**Pull request by NokeYuan**
inital commit for all codes

### CODE REVIEW

File: env.yaml [hunk 1]
```diff
@@ -0,0 +1,553 @@
...
+  - zeromq=4.3.4=hd77b12b_0
+  - zfp=0.5.5=hd77b12b_6
+  - zict=2.1.0=py39haa95532_0
+  - zipp=3.11.0=py39haa95532_0
+  - zlib=1.2.13=h8cc25b3_0
+  - zope=1.0=py39haa95532_1
+  - zope.interface=5.4.0=py39h2bbff1b_0
+  - zstandard=0.19.0=py39h2bbff1b_0
+  - zstd=1.5.2=h19a0ad4_0
+  - pip:
```
Comment (KevinHeUHN) [inline env.yaml:535]: Can you insert the contents of requirements.txt here under pip? And delete that file.

File: scripts/ed_association.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    train_results = load_results(train_results_path)
+
+    results_df = pd.DataFrame.from_dict(train_results, orient='index')
+    results_df.index.names = ['model', 'label']
+    results_df.reset_index(inplace=True)
+
+    test_df = pd.read_csv(test_df_path)
+    labels_list = [s for s in test_df.columns.tolist() if s.startswith("Label") and s.endswith("3pt_change")]
+
+    label_with_ed_per = process_results(results_df, test_df, labels_list, 'lgbm', 'Label_EDvisit')
```
Comment (KevinHeUHN) [inline scripts/ed_association.py:23]: add simple comments on what per and what all refers to

File: scripts/train_pipeline.py [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
+)
+from src.config import (
+    bounds_dict,
+    data_path_dict
+)
+
+from src.utility import get_neg_to_pos_ratios
+
+
+def perform_bayesian_optimization(objective_func, model_name, model_type, pbounds, train_df, tune_df,test_df, label,full_colmuns_list, n_iter=3, init_points=5, weights=None, return_optimizer=False,verbose = False,allow_duplicate_points = True,ci =False):
```
Comment (KevinHeUHN) [inline scripts/train_pipeline.py]: When functions have a lot of arguments, make it more readable by splitting the arguments into separate lines
```
def perform_bayesian_optimization(
    objective_func, 
    model_name, 
    model_type,
    ...
):
```

File: scripts/train_pipeline.py [hunk 4]
```diff
@@ -0,0 +1,463 @@
...
+
+    def objective_wrapper(**params):
+        nonlocal best_val_aucs  # Declare best_val_aucs as nonlocal so that we can modify it
+        nonlocal best_auroc  # Declare best_auroc as nonlocal so that we can modify it
+        nonlocal best_auroc_ci
+        nonlocal best_all_preds
+        nonlocal best_all_labels
+
+
+        objective_auroc, val_aucs ,auroc_ci, all_preds,all_labels,_,_ = objective_func(model_name, model_type, params, train_df, tune_df, test_df, label,full_colmuns_list, weights, verbose,ci)
```
Comment (KevinHeUHN) [inline scripts/train_pipeline.py]: arguments can also be split up into separate lines here too. In general, try not to go over 120 characters per line.

File: scripts/train_pipeline.py [hunk 5]
```diff
@@ -0,0 +1,463 @@
...
+        random_state=42,
+        verbose=2,
+        allow_duplicate_points=allow_duplicate_points
+    )
+
+    optimizer.maximize(n_iter=n_iter, init_points=init_points)
+
+    if return_optimizer:
+        # If return_optimizer is True, return a tuple containing both the optimizer and best_val_aucs
+        return optimizer, best_val_aucs,best_auroc_ci,best_all_preds,best_all_labels
```
Comment (KevinHeUHN) [inline scripts/train_pipeline.py:120]: When returning multiple outputs, I think it would be better to just a return a dictionary 
```
return {
    'optimizer': optimizer,
    'best_val_aucs': best_val_aucs,
    ...
}
```

File: scripts/train_pipeline.py [hunk 6]
```diff
@@ -0,0 +1,463 @@
...
+    else:
+        # Otherwise, just return the best parameters as before
+        return optimizer.max['params']
+
+def train_model(model_name,model_type,params, train_df,tune_df,test_df,label,full_colmuns_list,weights=None,verbose = False,ci= False,testing = False):
```
Comment (KevinHeUHN) [inline scripts/train_pipeline.py]: same here

File: scripts/train_pipeline.py [hunk 7]
```diff
@@ -0,0 +1,463 @@
...
+        # Testing
+        auroc_scores.append(roc_auc_score(all_labels, all_preds))
+        objective_auroc = np.mean(auroc_scores)
+
+        best_model = model
+        best_ir_model = isotonic
+
+
+    if model_type == 'dl':
+        torch.cuda.empty_cache()
```
Comment (KevinHeUHN) [inline scripts/train_pipeline.py:201]: redundant code? torch.cuda.empty_cache is already called at the top

File: scripts/train_pipeline.py [hunk 8]
```diff
@@ -0,0 +1,463 @@
...
+        learning_rate = params['learning_rate']
+        weight_decay = params['weight_decay']
+
+         # Check whether label and weights are given as list
+        if not isinstance(label, list):
+            label = [label]
+        if not isinstance(weights, list):
+            weights = [weights]
+
+        torch.cuda.empty_cache()
```
Comment (KevinHeUHN) [inline scripts/train_pipeline.py]: same here

File: scripts/train_pipeline.py [hunk 9]
```diff
@@ -0,0 +1,463 @@
...
+            num_channels = [int(params[f'num_channels_{i}']) for i in range(1, 4)]
+            kernel_size = int(params['kernel_size'])
+            # dropout = params['dropout']
+            model = TCN(input_size, num_channels, output_size, kernel_size=kernel_size,dropout=params['dropout']).to(device)
+
+        elif model_name == 'vanilla_transformer':
+            num_heads = {1: 8, 2: 16, 3: 32, 4: 64}
+            dim_feedforward = {1: 128, 2: 256, 3: 512}
+            hidden_size =  {1: 128, 2: 256, 3: 512}
+            model = Transformer(input_size, output_size,hidden_size =hidden_size[int(params['hidden_size'])],
```
Comment (KevinHeUHN) [inline scripts/train_pipeline.py:263]: nitpick but keep indentations consistent. Also try to clean up the spacing i.e. leave a space after a comma, don't have a space before =

File: scripts/train_pipeline.py [hunk 10]
```diff
@@ -0,0 +1,463 @@
...
+                                     num_heads=num_heads[int(params['num_heads'])],
+                                     num_layers=int(params['num_layers']),
+                                     dim_feedforward= dim_feedforward[int(params['dim_feedforward'])],dropout=params['dropout']).to(device)
+
+        optimizer = optim.Adam(model.parameters(), lr=learning_rate, weight_decay=weight_decay)
+
+        if testing == True:
+            train_losses, train_aucs, train_aups, val_losses, val_aucs, val_aups, mean_val_losses, train_auc_cis, train_aup_cis, val_auc_cis, val_aup_cis,ir_models,models,all_preds,all_labels = train_and_evaluate_dl(
```
Comment (KevinHeUHN) [inline scripts/train_pipeline.py]: As mentioned before, return a dictionary of these output, and separate a long line of arguments into separate lines

File: scripts/train_pipeline.py [hunk 11]
```diff
@@ -0,0 +1,463 @@
...
+        else:
+            train_losses, train_aucs, train_aups, val_losses, val_aucs, val_aups, mean_val_losses, train_auc_cis, train_aup_cis, val_auc_cis, val_aup_cis, ir_models, models, all_preds, all_labels = train_and_evaluate_dl(
+                30, model, train_loader, tune_loader, None, optimizer, criterion, device, output_size, label,weights,scheduler=None, early_stopping_patience=3, verbose=verbose, ci=ci, my_copy=False)
+
+        if output_size == 1:
+            best_epoch = np.argmax(val_aucs[0])
+            objective_auroc = val_aucs[0][best_epoch]  # find the epoch with the highest AUROC
+
+        else:
+            # Case when there are multiple outputs
```
Comment (KevinHeUHN) [inline scripts/train_pipeline.py:287]: No need to change anything now but why take the mean over all epochs? Instead of taking the highest auc across all epochs (for that particular target)?

The auc in the beginning would be bad since the model is just starting to learn.

File: scripts/train_pipeline.py [hunk 12]
```diff
@@ -0,0 +1,463 @@
...
+            search_results[(model, label)] = {
+                'params': optimizer.max['params'],
+                'target': optimizer.max['target'],  # Include the best objective value
+                'all_preds': all_preds[-1],
+                'all_labels': all_labels[-1]
+            }
+            all_evaluations[(model, label)] = optimizer.res
+
+        # Multi-task (all labels) for DL models
+        for model in dl_models:
```
Comment (KevinHeUHN) [inline scripts/train_pipeline.py:383]: Is this supposed to be inside the loop? (Do this for each label?)

File: src/train.py [hunk 13]
```diff
@@ -0,0 +1,451 @@
...
+                        f"Task {i + 1} {labels_list[i]}: Train AUC: {train_auc[-1]:.4f}, Train AUP: {train_aup[-1]:.4f}, Val Loss: {val_loss[-1]:.4f}, Val AUC: {val_auc[-1]:.4f}, Val AUP: {val_aup[-1]:.4f}")
+                print()
+            else:
+                print(f"Epoch {epoch + 1}/{epochs} - Train Loss: {train_loss:.4f}, Mean Val Loss: {mean_val_loss:.4f}")
+                for i, (train_auc, train_aup, train_auc_ci, train_aup_ci, val_loss, val_auc, val_aup, val_auc_ci,
+                        val_aup_ci) in enumerate(
+                        zip(train_aucs, train_aups, train_auc_cis, train_aup_cis, val_losses, val_aucs, val_aups,
+                            val_auc_cis, val_aup_cis)):
+                    print(
+                        f"Task {i + 1} {labels_list[i]}: Train AUC: {train_auc[-1]:.4f} (95% CI: {train_auc_ci[-1][0]:.4f}, {train_auc_ci[-1][1]:.4f}), Train AUP: {train_aup[-1]:.4f} (95% CI: {train_aup_ci[-1][0]:.4f}, {train_aup_ci[-1][1]:.4f}), Val Loss: {val_loss[-1]:.4f}, Val AUC: {val_auc[-1]:.4f} (95% CI: {val_auc_ci[-1][0]:.4f}, {val_auc_ci[-1][1]:.4f}), Val AUP: {val_aup[-1]:.4f} (95% CI: {val_aup_ci[-1][0]:.4f}, {val_aup_ci[-1][1]:.4f})")
```
Comment (KevinHeUHN) [inline src/train.py:425]: Long print statements can also be split into separate lines

File: src/train.py [hunk 14]
```diff
@@ -0,0 +1,451 @@
...
+
+        train_loss = train_dl(model, train_dataloader, optimizer, criterion, neg_to_pos_ratios, device)
+
+        # Ccalibrated model
+        if calibration == True:
+            ir_model_epoch = calibrated_model(model, tune_dataloader, neg_to_pos_ratios, output_size, device)
+        else:
+            ir_model_epoch = None
+
+        if my_copy == True:
```
Comment (KevinHeUHN) [inline src/train.py:365]: What is my_copy for?

File: src/utility.py
Comment (KevinHeUHN) [inline src/utility.py:1]: Nice. I think this can be split into different files. One for plots, called viz.py. One for evaluation functions, called eval.py. And the rest we should put in util.py, so we don't have two different utility files.

**Comment by KevinHeUHN** [comment]
Looks great! Only couple things to reorganize and nitpicks here and there.

**Review by KevinHeUHN — COMMENTED** [review]
Looks great! Only couple things to reorganize and nitpicks here and there.

File: src/train.py [hunk 15]
```diff
@@ -0,0 +1,451 @@
...
+def calibrated_model(model, dataloader, neg_to_pos_ratios,output_size, device):
+    model.eval()
+    running_loss = [0.0] * output_size
+    all_labels = [[] for _ in range(output_size)]
+    all_preds = [[] for _ in range(output_size)]
+
+    with torch.no_grad():
+        for i, batch in enumerate(dataloader):
+            inputs, labels = batch
+            inputs, labels = (inputs.to(device, dtype=torch.float), labels.to(device, dtype=torch.float))
```
Comment (KevinHeUHN) [inline src/train.py:145]: Also, try to minimize repeated code. validate_dl has the same block of code as here. You can place that block of code into a predict function.

**Review by KevinHeUHN — COMMENTED** [review]
Looks great! Only couple things to reorganize and nitpicks here and there.

File: scripts/train_pipeline.py (hunk 11 above)
Comment (NokeYuan) [reply scripts/train_pipeline.py:287]: `val_aucs` is a list of lists, initially designed for multi-task learning, where each sublist corresponds to the AUCs for a specific target. In train.py,` val_aucs` is defined as `val_aucs = [[] for _ in range(output_size)]`. If the output size is 1, this means` val_aucs` would have the structure` [[target1]]`, and `val_aucs[0] `would access the AUCs for this single target. Then, `best_epoch` selects the highest corresponding index of the list. 

In the multi-task scenarios, mean_aucs records the mean AUCs across 9 targets for each epoch. Then, using `np.argmax`, it finds the highest mean AUROC

File: scripts/train_pipeline.py (hunk 12 above)
Comment (NokeYuan) [reply scripts/train_pipeline.py:383]: Should be outside of the loop since it does not require loop over each target.  Fixed.

File: src/train.py (hunk 14 above)
Comment (NokeYuan) [reply src/train.py:365]: If copy is True, `copy.deepcopy(model)`  save the model(s) into list and return once finish training.

File: scripts/train_pipeline.py (hunk 11 above)
Comment (KevinHeUHN) [reply scripts/train_pipeline.py:287]: Ohhh I see I see. Yea makes sense, thanks for that explanation!

File: src/train.py (hunk 14 above)
Comment (KevinHeUHN) [reply src/train.py:365]: I see, so you have the option to store models for every epoch.

No need to change, but how I would do it is probably write checkpoint files for each epoch.

File: src/util.py [hunk 16]
```diff
@@ -4,8 +4,21 @@
 import os
 import pickle
 
+# Standard library imports
+import os
+from functools import reduce
+os.environ['OUTDATED_IGNORE'] = '1'
```
Comment (KevinJCHe) [inline src/util.py:10]: Just curious, what is this for?

File: src/util.py [hunk 17]
```diff
@@ -87,4 +100,237 @@ def get_excluded_numbers(df, mask: pd.Series, context: str = '.') -> None:
...
+            i += 1
+
+    if not interval_list:
+        return None
+
+    # Find the largest interval by comparing their lengths
+    largest_interval = max(interval_list, key=lambda x: x[1]-x[0])
+    return largest_interval
+
+def partial_predict(X,label,model_name,train_results,df,device):
```
Comment (KevinJCHe) [inline src/util.py:170]: Leave a quick docstring here explaining this is used for computing SHAP values for models taking sequential input (e.g. recurrent neural network)

**Review by KevinJCHe — CHANGES_REQUESTED** [review]
Yea nice job on creating the eval.py and viz.py!

Nothing too much to add from last time. Would you able to fix up the previous comments I left? 

Then we can merge this branch

File: src/util.py (hunk 16 above)
Comment (NokeYuan) [reply src/util.py:10]: it will ignore the outdated package reminder. E.g., xx latest version is 2.0.0 but your current is 1.5.3

Comment (KevinJCHe) [reply src/util.py:10]: Ahh cool thanks!

File: scripts/train_pipeline.py (hunk 4 above)
Comment (NokeYuan) [reply scripts/train_pipeline.py]: done

File: scripts/train_pipeline.py (hunk 6 above)
Comment (NokeYuan) [reply scripts/train_pipeline.py]: done

File: src/train.py (hunk 13 above)
Comment (NokeYuan) [reply src/train.py:425]: make sense, this is a print statament monitor the performance on the fly. Separate lines would casue taking up too much spaces on the screen, can comment out.

File: src/util.py (hunk 16 above)
Comment (NokeYuan) [reply src/util.py:10]: os.environ['OUTDATED_IGNORE'] = '1' would ignore message indicate which packages is outdated. E.g., xxx latest version is 2.0.0, your current version is 1.5.2

File: scripts/train_pipeline.py (hunk 3 above)
Comment (NokeYuan) [reply scripts/train_pipeline.py]: done, all functions that have greater than 5 arguments will be splitted into separate lines

File: src/util.py (hunk 17 above)
Comment (NokeYuan) [reply src/util.py:170]: done

File: scripts/ed_association.py (hunk 2 above)
Comment (NokeYuan) [reply scripts/ed_association.py:23]: done

File: env.yaml (hunk 1 above)
Comment (NokeYuan) [reply env.yaml:535]: done

**Review by KevinJCHe — APPROVED** [review]
